libraries/ESP8266WiFi/src/include/ClientContext.h

```diff
@@ -456,7 +456,7 @@ class ClientContext
             }
 
             _send_waiting = true;
-            esp_yield();
+            delay(_timeout_ms);
```

Comment:
This change may decrease bandwidth:
an ACK (= permission to write) could reach in before expiration of this delay.
I'll do some tests to show the impact.

Reply:
I expected to delay function to be returned sooner if esp_schedule() is called. The same patter is used in ClientContext::connect method where there is this comment:
// This delay will be interrupted by esp_schedule in the connect callback

I've tested and it seems that it works, esp_schedule seems to schedule cont task to continue immediately not waiting for full timer expiry.

Reply:
You seem to be right.
Was this WiFiClient change mandatory and the WiFiClientSecure change alone not sufficient according to you tests ?

Reply:
Is it always safe to call `delay()` from here?  Always guaranteed to be the right context?

The other thing is I don't think `delay(XX)` can legally return before the timer.  In all the OSes I've seen, `sleep` or `delay` guarantees a minimum time.  Otherwise, for example, if you were doing a SW PWM or something and a packet came in you'd have a short period.

Reply:
The function delay calls internally esp_yield which check right context using cont_can_yield condition. Thus, I think that we can call delay from any context. ESP8266 implementation seems to allow wakeup application context delay using esp_schedule. 

Currently delay doesn't guarantee that there is wait at full interval. If something use delay it is waken any time there is ack of packet or any other signal on esp_schedule. If you want proper timing, e.g. for generation precise output signal you should use approach like in Software serial where there is a loop based on ESP.getCycleCount():
#define WAIT { while (ESP.getCycleCount()-start < wait); wait += m_bitTime; }

The other option could be to put here delay(1) ms which will perform repeated looping until some packet is accepted and send buffer contains some free space or until connection timeout expires.

Reply:
@d-a-v Without this change esp_yield block execution until there is esp_schedule generated somewhere. In the case of unresponsive server, there are no acks and thus no esp_schedule in application having single WiFiClient connection. With multiple WiFiClient connections, write can be unblocked by ack on the other one but it has to come after that long timeout. In most cases acks will come in max few hundreds of ms (with delayed ACK enabled) and then nothing else will wake the currently blocked write after timeout expires. Other connections will not generate new data because (causing further acks) because application context is blocked here.